javascript/ql/src/semmle/javascript/MapAndSet.qll

```diff
@@ -0,0 +1,279 @@
+      exists(MapOrSetFlowStep step | step.loadStore(_, _, this, result))
+    }
+  }
+}
+
+/**
+ * A module for data-flow steps related to `Set` and `Map`.
+ */
+private module MapAndSetDataFlow {
+  private import MapsAndSets
+
+  /**
+   * A step for an `add` method, which adds an element to a Set.
+   */
+  private class SetAdd extends MapOrSetFlowStep, DataFlow::MethodCallNode {
+    SetAdd() { this.getMethodName() = "add" }
+
+    override predicate store(DataFlow::Node element, DataFlow::Node obj, string prop) {
+      obj = this.getReceiver().getALocalSource() and
+      element = this.getArgument(0) and
+      prop = setElement()
+    }
+  }
+
+  /**
+   * A step for the `Set` constructor, which copies any elements from the first argument into the resulting set.
+   */
+  private class SetConstructor extends MapOrSetFlowStep, DataFlow::NewNode {
+    SetConstructor() { this = DataFlow::globalVarRef("Set").getAnInstantiation() }
+
+    override predicate loadStore(
+      DataFlow::Node pred, DataFlow::Node succ, string fromProp, string toProp
+    ) {
+      pred = this.getArgument(0) and
+      succ = this and
+      fromProp = arrayLikeElement() and
+      toProp = setElement()
+    }
+  }
+
+  /**
+   * A step for a `for of` statement on a Map, Set, or Iterator.
+   * For Sets and iterators the l-value are the elements of the set/iterator.
+   * For Maps the l-value is a tuple containing a key and a value.
+   *
+   * This is partially duplicated behavior with the `for of` step for Arrays (in Arrays.qll).
+   * This duplication is required for the type-tracking steps defined in `MapsAndSetsTypeTracking`.
+   */
+  private class ForOfStep extends MapOrSetFlowStep, DataFlow::ValueNode {
+    ForOfStmt forOf;
+    DataFlow::Node element;
+
+    ForOfStep() {
+      this.asExpr() = forOf.getIterationDomain() and
+      element = DataFlow::lvalueNode(forOf.getLValue())
+    }
+
+    override predicate load(DataFlow::Node obj, DataFlow::Node e, string prop) {
+      obj = this and
+      e = element and
+      prop = arrayLikeElement()
+    }
+
+    override predicate loadStore(
+      DataFlow::Node pred, DataFlow::Node succ, string fromProp, string toProp
+    ) {
+      pred = this and
+      succ = element and
+      fromProp = mapValueUnknownKey() and
+      toProp = "1"
+    }
+  }
+
+  /**
+   * A step for a call to `forEach` on a Set or Map.
+   */
+  private class SetMapForEach extends MapOrSetFlowStep, DataFlow::MethodCallNode {
+    SetMapForEach() { this.getMethodName() = "forEach" }
+
+    override predicate load(DataFlow::Node obj, DataFlow::Node element, string prop) {
+      obj = this.getReceiver() and
+      element = this.getCallback(0).getParameter(0) and
+      prop = [setElement(), mapValueUnknownKey()]
+    }
+  }
+
+  /**
+   * A call to the `get` method on a Map.
+   * If the key of the call to `get` has a known string value, then only the value corresponding to that key will be retrieved.
```

Comment:
> If the key of the call to `get` has a known string value

This is confusing whe the type of `prop` in `load/3` is `string` - is that the known string? Can we perhaps use a `PseudoProperty`-type for the parameter instead?